### ISSUE TITLE
Adding Thermal Foundation in an existing world with Chromaticraft causes Resonant Ender recipes to produce Liquid Ender Essence instead

### ISSUE THREAD

**Issue by joakibj** [ReikaKalseki/Reika_Mods_Issues#686]
Added Thermal Foundation + Thermal Expansion to an existing world with Chromaticraft.

The Liquid Ender Essence does not get converted to Resonant Ender.

Wouldn't be an issue, except that it can't be used for crafting nor deposited or extracted from fluid containers. It can be imported into ExtraCells2, but not extracted. Thus it has very limited portability.

Any idea how to fix this?

**Comment by ReikaKalseki** [comment]
...Were you seriously expecting anything different? You created a world where the liquid "ender" is mapped to CC ender; why would it spontaneously remap, much less replace all the blocks in the world?

**Comment by joakibj** [comment]
Was not expecting the blocks in the world to be replaced, but rather a way for the Liquid Ender Essence to be converted to Resonant Ender (through the use of machines for example).

**Comment by 1skandar** [comment]
Wait one minute. While you are right, Reika in that this behavior is correct there is an issue here. 

joakibj: are you saying that since adding TE you can no longer pick up Liquid Ender Essence in buckets and the like? If so that sounds like an issue. 

Also, Reika there IS a bug here, regardless. I just tested this out of curiosity and adding TE to a world with Chromaticraft already in progress renders Enderium  blend uncraftable. The recipe is calling for Resonant Ender specifically and since Liquid Ender Essence is mapped to liquid "ender" there is no way to actually get Resonant Ender. Apparently that recipe is not respecting the fluid registry for whatever reason.

**Comment by ReikaKalseki** [comment]
This is because of that change Forge made with the Fluid registry some time ago (around build 1400) where fluid maps are now determined on a world-by-world basis and saved to it. Changing fluid IDs no longer causes an active remap.

And if the recipe is not respecting the registry, that is on their end.

**Comment by 1skandar** [comment]
Unless I'm missing something, once your Liquid Ender Essence is mapped as liquid "ender" adding TE should not be able to change that map, right? Thus, for that world, Liquid Ender Essence _is_ Resonant Ender for all intents and purposes? 

Which is where my request for clarification is coming in. He said " it can't be used for crafting nor deposited or extracted from fluid containers" and if he is talking about Liquid Ender Essence there something has gone askew. In that case it would mean a re-map has occured, to Resonant Ender at a guess, leaving Liquid Ender Essence a fluid without an associated fluid registry and thus rendering it useless. And, again, that shouldn't happen, right? A new mod's fluid should not be able to supplant an old mod's fluid on the registry for exactly this reason?  And if it has something has gone wrong?

**Comment by joakibj** [comment]
@1skandar : Liquid Ender Essence can still be picked up with buckets. They can be deposited from the bucket to the world. But there ends its interaction with the world. A bucket of Liquid Ender Essence can not be emptied into a fluid container of any sort. A Spring-powered pump with Liquid Ender Essence in it is useless. An ExtraCells2 Fluid ME Cell with Liqid Ender Essence in it can not be exported.

Enderium Blend can be crafted if the world had for example a reservoir with Liquid Ender Essence in it. It can be pumped into the Fluid Transposer and that will create buckets of Resonant Ender. But with pure manual interaction it is a useless fluid.

My intention with this issue was to document this occurence for others and possibly find a solution. I.e. a block or some game mechanism to convert from Liquid Ender Essence to Resonant Ender. I had no allusions as to the fluid magically being the same or replace blocks in the world.

**Comment by 1skandar** [comment]
Aha, ok I see it.

Hey, Reika I _think_ there may actually be a problem here. Your Liquid Ender Essence cannot be used in a Liquid Transposer or a Portable Tank just from my limited testing with just Chromaticraft and Thermal Expansion despite being correctly mapped to the fluid registry. 

Edit: Added Extra Utilities and Buildcraft. Neither drums nor Buildcraft tanks can be used either.

**Comment by 1skandar** [comment]
Confirmed, there is a problem here, and I'm not sure on whose end. 

What I did:
Made an MultiMC instance with:
DragonAPI
Chromaticraft
NEI 
NEI Integration

Created a world, then dumped the fluid registry. Verified Liquid Ender Essence was correctly mapped. Exited Minecraft.

Added COFH and all three TE mods. 

Re-entered world. Verified Liquid Ender Essence was still correctly mapped. Tested Liquid Ender Essence via bucket (passed) portable tank (fail) liquid transposer (fail). Exited Minecraft.

Added Buildcraft and Extra Utilities (I needed other liquid Containers and  those mods were handy)

Entered Minecraft. Tested Buildcraft Tank (fail) and Extra Utility drum (fail) Exit Minecraft.

Remove COFH Core and all TE mods. 

Enter Minecraft. Test Buildcraft tank (success) and Extra Utilities drum (also success)

So, to sum up, it looks like there is some sort of cross mod interaction with TE going on, but I couldn't _begin_ to tell you what. As soon as you add TE, the **only** container the Liquid Ender Essence is useful in is a bucket, it immediately gets removed from all others and, apparently, ALL crafting recipes **despite** still being mapped to the fluid registry.

**Comment by ReikaKalseki** [comment]
Something is overriding the mappings. Either Forge's fluid persistence was only half-implemented and all containers and recipes still get remapped to the new object (under the old "whatever gets registered first" behavior), leaving them useless - which would be **monumentally** stupid - or it is a deliberate, direct (probably reflective) insertion into the maps. The former seems more likely - especially knowing that TF fluids are registered extremely early - but either way someone has done something _very_ wrong.

Also, liquid ender essence is supposed to be entirely interchangeable with liquid ender even without the fluid registry; it is after all a relic of the 1.5.2 Ender Forest days when its explicit purpose was to make things like RC circuit boards (which lacked the silicon and electrum versions) and Tesserracts cheaper. What this then means is that I am entirely willing to add some sort of conversion system - up to and including in-world replacement and some sort of "detection and conversion" for fluid tiles and containers - provided it can be done safely and performantly.

**Comment by 1skandar** [comment]
Well, check the thread on COFH's side. They claim it isn't them. 

At any rate, I've done my due diligence. You now have a clearly reproducible bug. And one I don't envy you sorting out. Luckily.this issue should be vanishingly rare.

**Comment by ReikaKalseki** [comment]
I will see what I can do about a converter.
